cpp/src/lists/explode.cu

```diff
@@ -208,38 +208,45 @@ std::unique_ptr<table> explode_outer(table_view const& input_table,
   // offsets + 1 here to skip the 0th offset, which removes a - 1 operation later.
   auto offsets_minus_one = thrust::make_transform_iterator(
     thrust::next(offsets), [offsets] __device__(auto i) { return (i - offsets[0]) - 1; });
+
+  // we need to do this loop at least explode_col times or we may not properly fill in null and
+  // empty entries.
+  auto loop_count = std::max(sliced_child.size(), explode_col.size());
+
   // Fill in gather map with all the child column's entries
   thrust::for_each(rmm::exec_policy(stream),
                    counting_iter,
-                   counting_iter + sliced_child.size(),
+                   counting_iter + loop_count,
                    [offsets_minus_one,
-                    gather_map             = gather_map.begin(),
-                    explode_col_gather_map = explode_col_gather_map.begin(),
-                    position_array         = pos.begin(),
+                    gather_map_p             = gather_map.begin(),
+                    explode_col_gather_map_p = explode_col_gather_map.begin(),
+                    position_array           = pos.begin(),
+                    sliced_child_size        = sliced_child.size(),
+                    null_or_empty_offset_p   = null_or_empty_offset.begin(),
                     include_position,
                     offsets,
-                    null_or_empty_offset = null_or_empty_offset.begin(),
                     null_or_empty,
                     offset_size = explode_col.offsets().size() - 1] __device__(auto idx) {
```

Comment:
You call a lambda which also calls `thrust::distance` and `thrust::lower_bound`, which in all result in a huge loop. Can we extract the lambda into a functor?

Reply:
I'm not sure what that would buy us here. What goal are you hoping to achieve by doing this? I'm not against it, just trying to undertsand.

Reply:
Or just before the algorithm call, define the lambda and store it in a variable `auto func = [...]__device__( auto idx) {...}`